src/vsr/replica.zig

```diff
@@ -1939,6 +2008,203 @@ pub fn ReplicaType(
+                        requests.len - i,
+                        requests.len,
+                    });
+                    return;
+                };
+
+                log.debug("{}: on_request_blocks: reading block " ++
+                    "(replica={} address={} checksum={})", .{
+                    self.replica,
+                    message.header.replica,
+                    request.block_address,
+                    request.block_checksum,
+                });
+
+                read.* = .{
+                    .replica = self,
+                    .destination = message.header.replica,
+                    .read = undefined,
+                };
+
+                self.grid.read_block_repair(
+                    on_block_read_repair,
+                    &read.read,
+                    request.block_address,
+                    request.block_checksum,
+                );
+            }
+        }
+
+        fn on_block(self: *Self, message: *const Message) void {
+            maybe(!self.state_machine_opened);
+            assert(message.header.command == .block);
+            assert(message.header.size <= constants.block_size);
+            assert(message.header.op > 0); // op holds the block's address.
+
+            // TODO State sync
+
+            if (!self.grid.faulty(message.header.op, message.header.checksum)) {
+                log.debug("{}: on_block: ignoring; block not repairing (address={} checksum={})", .{
+                    self.replica,
+                    message.header.op,
+                    message.header.checksum,
+                });
+                return;
+            }
+
+            switch (self.grid.writing(message.header.op, null)) {
+                .init => unreachable,
+                .repair => {
+                    log.debug("{}: on_block: ignoring; already writing block (address={} checksum={})", .{
+                        self.replica,
+                        message.header.op,
+                        message.header.checksum,
+                    });
+                    return;
+                },
+                .none => {},
+            }
+
+            const write = self.grid_writes.acquire() orelse {
+                log.debug("{}: on_block: ignoring; no write available (address={} checksum={})", .{
+                    self.replica,
+                    message.header.op,
+                    message.header.checksum,
+                });
+                return;
+            };
+
+            log.debug("{}: on_block: repairing block (address={} checksum={})", .{
+                self.replica,
+                message.header.op,
+                message.header.checksum,
+            });
+
+            write.* = .{
+                .replica = self,
+                .write = undefined,
+            };
+
+            const block = &self.grid_write_blocks[self.grid_writes.index(write)];
+            stdx.copy_disjoint(.inexact, u8, block.*, message.buffer[0..message.header.size]);
```

Comment:
:thinking: so we have a memcpy on both read and write path here. This feels like a relatively big deal, as block repair is our most perf-sensitive operation (when we repair, we might have a lot of data to transfer). Why do we not use our pre-allocated message pool here?